plugins/ingest/src/main/java/org/elasticsearch/ingest/Data.java

```diff
@@ -47,67 +48,95 @@ public Data(Data other) {
         this(other.index, other.type, other.id, new HashMap<>(other.document));
     }
 
-    @SuppressWarnings("unchecked")
-    public <T> T getProperty(String path) {
-        // TODO: we should not rely on any core class, so we should have custom map extract value logic:
-        // also XContentMapValues has no support to get specific values from arrays, see: https://github.com/elastic/elasticsearch/issues/14324
-        return (T) XContentMapValues.extractValue(path, document);
+    /**
+     * Returns the value contained in the document for the provided path
+     * @param path The path within the document in dot-notation
+     * @param clazz The expected class of the field value
+     * @return the value for the provided path if existing, null otherwise
+     * @throws IllegalArgumentException if the field is present but is not of the type provided as argument.
+     */
+    public <T> T getProperty(String path, Class<T> clazz) {
+        Object property = get(path);
```

Comment:
I was also thinking, shall we reject the property suffix and make it getValue? just a thought...